### PR TITLE
[WPE][GTK] Remove webkit_web_context_set_sandbox_enabled() from modern API

### PR DESCRIPTION
**Pull request by mcatanzaro**
#### 0f14b00d81e3b31d47e232dab679a21391d9d0b2
<pre>
[WPE][GTK] Remove webkit_web_context_set_sandbox_enabled() from modern API
<a href="https://bugs.webkit.org/show_bug.cgi?id=220117">https://bugs.webkit.org/show_bug.cgi?id=220117</a>

Reviewed by Adrian Perez de Castro.

* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
(webkitWebContextConstructed):
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.h.in:
* Source/WebKit/gtk/migrating-to-webkitgtk-6.0.md:
* Tools/MiniBrowser/gtk/main.c:
(activate):
* Tools/PlatformWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/259028@main">https://commits.webkit.org/259028@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/43b1bb4a455e8d47945b99a8bf9aa2862432a2de

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12770 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112883 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173214 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107606 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3667 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95895 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112039 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109426 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93698 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38355 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25302 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80006 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6138 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26702 "Found 1 new test failure: media/video-playback-quality.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6315 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3221 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12297 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46216 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6201 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8072 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by mcatanzaro** [comment]
Depends on #5943

File: Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp [hunk 1]
```diff
@@ -1353,6 +1378,9 @@ static bool pathIsBlocked(const char* path)
     if (!g_path_is_absolute(path))
         return true;
 
+    if (pathIsUserHomeDirectory(path))
+        return true;
+
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp]: This is unrelated to this patch, no? I would move it to its own bug and patch.

File: Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp [hunk 2]
```diff
@@ -187,11 +187,15 @@ void ProcessLauncher::launchProcess()
...
-    if (sandboxEnv)
-        sandboxEnabled = !strcmp(sandboxEnv, "1");
+    const char* sandboxEnv = g_getenv("WEBKIT_FORCE_SANDBOX");
+    if (sandboxEnv) {
+        if (!strcmp(sandboxEnv, "1"))
+            sandboxEnabled = true;
+        else
+            g_warning("WEBKIT_FORCE_SANDBOX no longer allows disabling the sandbox. Use WEBKIT_ALLOW_HACKING_ME=1 instead.");
+    } else if ((sandboxEnv = g_getenv("WEBKIT_ALLOW_HACKING_ME")))
+        sandboxEnabled = strcmp(sandboxEnv, "1");
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: Why do we keep the force sandbox env var then? `WEBKIT_ALLOW_HACKING_ME` doesn't sound like a good name either. I would simply use `WEBKIT_DISABLE_SANDBOX` which is useful, not for being hacked, but for debugging purposes.

**Review by carlosgcampos — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp (hunk 1 above)
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp]: OK.

File: Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp (hunk 2 above)
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: > Why do we keep the force sandbox env var then? 

(a) It can be used to force enable the sandbox. (b) It's useful to provide a warning to people who were previously expecting it to work.

> I would simply use WEBKIT_DISABLE_SANDBOX which is useful, not for being hacked, but for debugging purposes.

This doesn't really convey the badness of disabling the sandbox. I've seen _many_ users running with this environment variable as a workaround for various historical problems without realizing how risky it is to do so.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: CC @TingPing for opinion

Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: > (a) It can be used to force enable the sandbox. (b) It's useful to provide a warning to people who were previously expecting it to work.

why do we need to force the sandbox if it's enabled by default? I guess we only want to keep it for gtk3 builds, env vars are not public api, we don't need to keep compatibility and we are bumping the api version in any case.

> This doesn't really convey the badness of disabling the sandbox. I've seen many users running with this environment variable as a workaround for various historical problems without realizing how risky it is to do so.

Add a g_warning if you want, but the name of the env var should be something like `WEBKIT_DISABLE_SANDBOX` which is exactly what it does.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: > why do we need to force the sandbox if it's enabled by default? I guess we only want to keep it for gtk3 builds, env vars are not public api, we don't need to keep compatibility and we are bumping the api version in any case.

Oh yeah, that is true.

> Add a g_warning if you want, but the name of the env var should be something like WEBKIT_DISABLE_SANDBOX which is exactly what it does.

Let's just not have the environment variable for now until we can agree on the name. It seems useful for debugging, but if it's not named with care, then it is dangerous.

**Comment by mcatanzaro** [comment]
Unfortunately, cog is already using the new API version:

```
[34/37] Linking target launcher/cog
FAILED: launcher/cog 
cc  -o launcher/cog launcher/cog.p/cog.c.o launcher/cog.p/cog-launcher.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 '-Wl,-rpath,$ORIGIN/../core:/app/webkit/WebKitBuild/Release/lib:/usr/lib/x86_64-linux-gnu' -Wl,-rpath-link,/app/webkit/WebKitBuild/Release/Tools/cog-prefix/src/cog-build/core -Wl,-rpath-link,/app/webkit/WebKitBuild/Release/lib -Wl,-rpath-link,/usr/lib/x86_64-linux-gnu -Wl,--start-group core/libcogcore.so.9.2.0 /app/webkit/WebKitBuild/Release/lib/libWPEWebKit-2.0.so /usr/lib/x86_64-linux-gnu/libsoup-3.0.so -Wl,--export-dynamic /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so -pthread /usr/lib/x86_64-linux-gnu/libglib-2.0.so /usr/lib/x86_64-linux-gnu/libgio-2.0.so /usr/lib/x86_64-linux-gnu/libgobject-2.0.so /usr/lib/x86_64-linux-gnu/libwpe-1.0.so /usr/lib/x86_64-linux-gnu/libmanette-0.2.so -Wl,--end-group
/usr/lib/gcc/x86_64-unknown-linux-gnu/12.1.0/../../../../x86_64-unknown-linux-gnu/bin/ld: launcher/cog.p/cog-launcher.c.o: in function `cog_launcher_startup':
cog-launcher.c:(.text+0x1b74): undefined reference to `webkit_web_context_set_sandbox_enabled'
collect2: error: ld returned 1 exit status
```

If we want to keep cog on the wpewebkit-2.0 API version, then it's a chicken/egg problem: the WebKit CI won't pass unless we update cog first, but cog's CI cannot pass unless we update WebKit first. Can we switch cog back to wpewebkit-1.1 for now?

File: Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp (hunk 2 above)
Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: `WEBKIT_ALLOW_HACKING_ME` is a bit over the top...

If you really want to be verbose I agree it should at least be descriptive, like `WEBKIT_DISABLE_SECURITY_FEATURE_SANDBOX` or something. That is pretty loud but also clear.

**Comment by mcatanzaro** [comment]
I'll return to this next year. Depends on #7508.

**Comment by mcatanzaro** [comment]
> I'll return to this next year. Depends on #7508.

Well, actually it now instead depends on #8011

File: Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp (hunk 2 above)
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: Let's use `WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS`. A bit longer, but hopefully that's fine.

**Comment by mcatanzaro** [comment]
Updated. Note this also fixes https://bugs.webkit.org/show_bug.cgi?id=249034

File: Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp (hunk 2 above)
Comment (carlosgcampos) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: This is missing in the patch in the end, no? I still don't like the name, but I won't block this anymore either.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp]: Yes, I will split it out to https://bugs.webkit.org/show_bug.cgi?id=250232 because you asked me to do it in a different bug. Feel free to suggest names there.

File: Tools/PlatformWPE.cmake [hunk 3]
```diff
@@ -27,10 +27,11 @@ if (ENABLE_COG)
         set(WPE_COG_PKG_CONFIG_PATH ${CMAKE_BINARY_DIR})
     endif ()
     if ("${WPE_COG_REPO}" STREQUAL "")
-        set(WPE_COG_REPO "https://github.com/Igalia/cog.git")
+        # FIXME: Switch back to the upstream repo https://github.com/Igalia/cog.git.
+        set(WPE_COG_REPO "https://github.com/mcatanzaro/cog.git")
```
Comment (clopez) [inline Tools/PlatformWPE.cmake]: I'm not very happy with the idea of changing the Cog upstream repo here.
Can we find a way of landing this on cog upstream disabled by default? so we can get Cog building both with the old API and also with the newer one? An idea to do that is that we can pass here to the Cog build system an optional build parameter similar to the one used here: `-DENABLE_2022_GLIB_API` that will allow in Cog to select using the newer api or the previous one.

Comment (mcatanzaro) [reply Tools/PlatformWPE.cmake]: Does cog really need to support three different WPE API versions? Is it really a valuable thing to do? Yes it's certainly possible to do, but I figured two APIs would probably be enough? I did not remove the support for wpewebkit-1.0 and libsoup 2. I only removed wpewebkit-1.1 (libsoup 3 with old API). It's really not clear to me that we need to keep wpewebkit-1.1 in WebKit either; I suspect nothing uses it and it's just hanging around still because nobody has made the decision to remove it yet?

> I'm not very happy with the idea of changing the Cog upstream repo here.

If you give me access to Igalia/cog.git then I could land commits in a side branch there instead of having to use a fork. Alternatively, you could accept https://github.com/Igalia/cog/pull/524, but this would require some work to make the cog CI pass. Another alternative: retarget that pull request for a side branch and accept it into the side branch, so we can merge it to the upstream repo now, but in a side branch where we don't worry about fixing the CI. Then there's no longer any need to switch to a fork repo here and WebKit development can be unblocked.

BTW, we are running _very_ low on time to complete the API changes, so I don't think it makes sense to block on cog any longer tbh. I suggest we should return to cog next month _after_ API freeze begins on Feb 11. (We'll probably run past that deadline, but should try to get as much done before then as possible.)

Comment (carlosgcampos) [reply Tools/PlatformWPE.cmake]: I guess we can also use a branch in cog repo and anybody can create a pr targeting that branch, no?

Comment (mcatanzaro) [reply Tools/PlatformWPE.cmake]: > I guess we can also use a branch in cog repo and anybody can create a pr targeting that branch, no?

Yes. But I can't do it myself because I don't have permission. So please either do it, or give me permission so we can unblock this.

Comment (mcatanzaro) [reply Tools/PlatformWPE.cmake]: Patrick created https://github.com/Igalia/cog/tree/mcatanzaro/next so that this pull request can proceed without changing the origin repo.

**Comment by mcatanzaro** [comment]
Now it looks like cmake is not refreshing the cog repo to notice the new commit? I will have to do a WPE build tomorrow because it seems to be generating a cmake file in the build directory, and that's the file I'll need to debug....

**Review by aperezdc — APPROVED** [review]
